tools/c7n_azure/c7n_azure/container_host/modes.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2019 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+
+from c7n_azure.constants import (CONTAINER_EVENT_TRIGGER_MODE,
+                                 CONTAINER_TIME_TRIGGER_MODE)
+from c7n_azure.policy import AzureModeCommon
+
+from c7n import utils
+from c7n.policy import PullMode, ServerlessExecutionMode, execution
+
+
+class AzureContainerHostMode(ServerlessExecutionMode):
```

Comment:
I'm a little unclear why this modules is needed. event modes in other providers, have different entry points then run. ie. this feels like a lot boilerplate. effectively this wants to run all the event mode policies, having all of those switched out by manually switching out config, rather than just running the daemon on the policy set, feels odd.

Reply:
I think you are basically saying that these modes are overkill because we could just run the bits right in the daemon.  So two things - 

- I want the policies to run OK at the normal custodian CLI because it makes dev really hard otherwise (this may actually be a more core feature request solved with a mode override flag).
- I am not 100% sure that I don't want to support any provisioning code yet (although I'm increasingly convinced it isn't needed).  

It may be that I am currently thinking of this as both an alternative to serverless and an alternative to the CLI (e.g. Functions compete and a daemon mode) - when in reality I need to pick one direction so as to avoid it being a mess...

Reply:
the concern is that your multiplying modes for a different execution, which makes becomes unwieldy as the number of modes increases and causes a lot of policy change churn to try out the new execution mode.

running  policies on the cli doesn't really have meaning for an event mode, except via dryrun unless its just provisioning the resources, because there is no event per se.

Reply:
Yes I agree it is a problem, just not sure how to solve it without some larger core improvements which allow decoupling global config from policies.

IMO the mode shouldn't have deployment details in it because those make policy maintenance/authoring challenging, then we can use generic event/schedule modes.

Are you recommending I modify the schema to be a `OneOf` type thing based on Hosting Type? So if its Azure Functions we require a bunch of different fields and run different provisioning code, and if its container host we have the simpler set of fields?  This is definitely an option I'm open to - was just worried those modes and schemas would get too complicated.

re: dev/event mode - It might be different for AWS events, but events for us are basically the same result as just adding a value filter with a resource ID, so in order to write those policies we always just comment out the mode header to dev them with the CLI, then remove the comment block before check-in.  For periodic of course there is no reason not to run at CLI.  Some flag (maybe even supporting a resource ID to emulate an event) might make this nicer.  Anyway, this is a future feature request :)